src/lib/storage/segment_access_counter.cpp

```diff
@@ -0,0 +1,131 @@
+#include "segment_access_counter.hpp"
+
+#include <numeric>
+#include <sstream>
+
+#include "boost/format.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+SegmentAccessCounter::SegmentAccessCounter() {
+  DebugAssert((size_t)AccessType::Count == access_type_string_mapping.size(),
+    "access_type_string_mapping should contain as many entries as there are access types.");
+  reset();
+}
+
+SegmentAccessCounter::SegmentAccessCounter(const SegmentAccessCounter& counter) {
+  _set_counters(counter);
+}
+
+SegmentAccessCounter& SegmentAccessCounter::operator=(const SegmentAccessCounter& counter) {
+  _set_counters(counter);
+  return *this;
+}
+
+void SegmentAccessCounter::_set_counters(const SegmentAccessCounter& counter) {
+  for (auto counter_index = 0ul, size = _counters.size(); counter_index < size; ++counter_index) {
+    _counters[counter_index] = counter._counters[counter_index].load();
+  }
+}
+
+SegmentAccessCounter::CounterType& SegmentAccessCounter::get(const AccessType type) {
+  return _counters[(size_t)type];
+}
+
+const SegmentAccessCounter::CounterType& SegmentAccessCounter::get(const AccessType type) const {
+  return _counters[(size_t)type];
+}
+
+uint64_t SegmentAccessCounter::sum() const {
+  return std::accumulate(_counters.cbegin(), _counters.cend() - 1, 0ul);
+}
+
+std::string SegmentAccessCounter::to_string() const {
+  std::string result = std::to_string(_counters[0]);
+  result.reserve((size_t)AccessType::Count * 19);
+  for (auto access_type = 1u; access_type < (size_t)AccessType::Count; ++access_type) {
+    result.append(",");
+    result.append(std::to_string(_counters[access_type]));
+  }
+  return result;
+}
+
+// Computes the AccessType from the given positions list by determining the access pattern
+SegmentAccessCounter::AccessType SegmentAccessCounter::access_type(const PosList& positions){
+  const auto access_pattern = _access_pattern(positions);
+  switch (access_pattern) {
+    case SegmentAccessCounter::AccessPattern::Point:
+      return AccessType::Point;
+    case SegmentAccessCounter::AccessPattern::SequentiallyIncreasing:
+    case SegmentAccessCounter::AccessPattern::SequentiallyDecreasing:
+      return AccessType::Sequential;
+    case SegmentAccessCounter::AccessPattern::RandomlyIncreasing:
+    case SegmentAccessCounter::AccessPattern::RandomlyDecreasing:
+      return AccessType::Increasing;
+    case SegmentAccessCounter::AccessPattern::Random:
+      return AccessType::Random;
+  }
+  Fail("This code should never be reached.");
+}
+
+void SegmentAccessCounter::reset() {
+  for (auto& counter : _counters) counter = 0ul;
+}
+
+// Sets all counters of all segments in tables to zero.
+void SegmentAccessCounter::reset(const std::map<std::string, std::shared_ptr<Table>>& tables) {
+  for (const auto& [table_name, table_ptr] : tables) {
```

Comment:
Just `table` (instead of `table_ptr`) would be more consistent with the rest of the code base; same with chunk.